sky/provision/aws/instance.py

```diff
@@ -0,0 +1,132 @@
+"""AWS instance provisioning."""
+from typing import Dict, List, Any, Optional
+
+from sky.provision import common
+from sky.provision.aws import utils
+
+# ======================== About AWS subnet/VPC ========================
+# https://stackoverflow.com/questions/37407492/are-there-differences-in-networking-performance-if-ec2-instances-are-in-differen
+# https://docs.aws.amazon.com/vpc/latest/userguide/how-it-works.html
+# https://docs.aws.amazon.com/vpc/latest/userguide/configure-subnets.html
+
+# ======================== Instance state and lifecycle ========================
+# https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html
+
+# ======================== About AWS availability zone ========================
+# Data transfer within the same region but different availability zone
+#  costs $0.01/GB:
+# https://aws.amazon.com/ec2/pricing/on-demand/#Data_Transfer_within_the_same_AWS_Region
+
+# Tag uniquely identifying all nodes of a cluster
+TAG_RAY_CLUSTER_NAME = 'ray-cluster-name'
+
+
+def _filter_instances(ec2, filters: List[Dict[str, Any]],
+                      included_instances: Optional[List[str]],
+                      excluded_instances: Optional[List[str]]):
+    instances = ec2.instances.filter(Filters=filters)
+    if included_instances is not None and excluded_instances is not None:
+        raise ValueError('"included_instances" and "exclude_instances"'
+                         'cannot be specified at the same time.')
+    if included_instances is not None:
+        instances = instances.filter(InstanceIds=included_instances)
+    elif excluded_instances is not None:
+        included_instances = []
+        for inst in list(instances):
+            if inst.id not in excluded_instances:
+                included_instances.append(inst.id)
+        instances = instances.filter(InstanceIds=included_instances)
+    return instances
+
+
+def stop_instances(region: str,
+                   cluster_name: str,
+                   included_instances: Optional[List[str]] = None,
+                   excluded_instances: Optional[List[str]] = None) -> None:
+    """See sky/provision/__init__.py"""
+    ec2 = utils.create_resource('ec2', region=region)
+    filters = [
+        {
+            'Name': 'instance-state-name',
+            'Values': ['pending', 'running'],
+        },
+        {
+            'Name': f'tag:{TAG_RAY_CLUSTER_NAME}',
+            'Values': [cluster_name],
+        },
```

Comment:
For this `ray` specific filtering, I am thinking maybe we can move it upper to the `backend.provision_lib` and have this `stop_instances` having the following signature:
```
def stop_instances(region: str,
                   cluster_name: str,
                   tag_filters: Dict[str, str],
                   excluded_instances: Optional[List[str]] = None) -> None:
```
The include/exclude may also be done with the `tag_filters`, just as what we did for the `cloud.query_status`?

Reason for this change: I am wondering if we can make the `provision` module decoupled with the ray-specific implementation. All the ray backend related things can go under `backends/`. Wdyt?